lightning/src/offers/mod.rs

```diff
@@ -0,0 +1,231 @@
+// This file is Copyright its original authors, visible in version control
+// history.
+//
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE
+// or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! Implementation of Lightning Offers
+//! ([BOLT 12](https://github.com/lightning/bolts/blob/master/12-offer-encoding.md)).
+
+use bitcoin::bech32;
+use bitcoin::bech32::FromBase32;
+use bitcoin::hash_types::BlockHash;
+use bitcoin::secp256k1::{PublicKey, XOnlyPublicKey};
+use bitcoin::secp256k1::schnorr::Signature;
+use core::str::FromStr;
+use ln::PaymentHash;
+use ln::features::OfferFeatures;
+use ln::msgs::DecodeError;
+use util::ser::{HighZeroBytesDroppedVarInt, Readable, WithLength, WithoutLength};
+
+use prelude::*;
+
+/// An `offer` TLV stream without any semantic checks, apart from any checks performed when parsing
```

Comment:
> That's a separate issue though? If we want more errors out of the decoder, we can create those, lets not avoid using our decoder just because we want to add more features to the decoder.

I guess this is related to discussion below about code organization. If `OfferTlvStream` is thought to be an underlying implementation of `Offer`, then there are a few semantic checks related to field presence that would be required:
- `description` is present
- either `node_id` or `paths` is present but not both
- `refund_for` is present only if `send_invoice` is present

I wouldn't want to split these semantic checks across two layers. But if there is only an `Offer` struct, then all these checks would be written in a custom `Readable`, IIUC, where the first is performed by a `read_tlv_fields!` and the other two are handwritten.

> I'm generally very much against this kind of approach - I'm strongly in the "write your datastructures right and the code writes itself" camp - the datastructures should thus generally not be able to represent an invalid state to the greatest extent possible. That implies doing checks during deserialization and refusing to create an object that is invalid, not deserializing and then checking later, leading to tons of unwraps in the accessors assuming the preconditions were checked correctly.

Hmmm... I think there's a tradeoff here. The code won't write itself in this case because we are required to adhere to the encoding as defined in BOLT 12, which doesn't have concepts of enums or structs with optional values, for instance. So the resulting `Readable` and `Writeable` implementations would be very procedural and prone to have inconsistencies.

On the other hand, the `OfferTlvStream` approach leads to a very declarative format, semantic checks are separated from type checks, and serialization is free and therefore less error prone (i.e., doesn't require deconstructing types into component TLV records). Given only one field is required, there should be only one unwrap; the optional field accessors are mostly of the nature: `self.tlv_stream.field.as_ref()`. Additionally, the wrapper `Offer` struct could only be created from a semantically valid `OfferTlvStream` since `Offer`'s decoder would be responsible for performing semantic checks on `OfferTlvStream`. The drawback is accessors for complex types (i.e., enums, structs) may need to access more than one TLV record.

Reply:
Will push a couple WIP commits showing the accessor API with wrapped types and separated semantic checks. I think I'd prefer to at least have the intermediary struct to enforce the TLV encoding, even if forgoing the wrapper approach. It gives a much better separation between type checks and semantic checsks. The `OfferTlvStream` can always be deconstructed into parts when converting to `Offer`. There's nothing performance critical here.